### ISSUE TITLE
lsp-mode hangs while "starting" over tramp without lsp-print-io

### ISSUE THREAD

**Issue by shanemcd** [emacs-lsp/lsp-mode#2709]
Hello. I have been debugging hanging / performance issues with lsp-mode when I stumbled upon: https://emacs-lsp.github.io/lsp-mode/page/performance/#check-if-logging-is-switched-off

I tried to do this, but now lsp-mode does not start at all. Adding back `(setq lsp-log-io t)` solves the "starting" issue, but with that I start to see emacs freezing after using it for some time.

I found https://github.com/emacs-lsp/lsp-mode/issues/1845, which indicated that updating tramp helped, but I am still seeing the issue after updating to tramp 2.5.0.2.

**To Reproduce**
I am able to reproduce this when using any Python project over Tramp. Here is my complete lsp-mode / lsp-ui config:

```lisp
(use-package lsp-mode
  :init (setq lsp-keymap-prefix "s-l")
  :hook ((python-mode . lsp))
  :commands (lsp lsp-deferred)
  :config
  ;; (setq lsp-log-io t)
  (setq lsp-pyls-configuration-sources ["flake8"])
  (setq lsp-enable-file-watchers nil)
  (setq lsp-pyls-plugins-mccabe-enabled nil)
  (setq gc-cons-threshold 100000000)
  (setq read-process-output-max (* 1024 1024)) ;; 1mb
  (setq lsp-diagnostic-package :none)
  (lsp-register-client
    (make-lsp-client :new-connection (lsp-tramp-connection "pyls")
                     :major-modes '(python-mode)
                     :remote? t
                     :server-id 'pyls-remote)))

(use-package lsp-ui
  :requires lsp-mode flycheck
  :commands lsp-ui-mode
  :custom-face
  (lsp-ui-doc-background ((t (:background nil))))
  (lsp-ui-doc-header ((t (:inherit (font-lock-string-face italic)))))
  :hook (lsp-mode-hook . lsp-ui-mode)
  :custom
  (lsp-ui-doc-enable t)
  (lsp-ui-doc-use-childframe f)
  (lsp-ui-flycheck-enable t)
  (lsp-ui-flycheck-list-position 'right)
  (lsp-ui-flycheck-live-reporting t)
  (lsp-ui-peek-enable t)
  (lsp-ui-peek-list-width 60)
  (lsp-ui-peek-peek-height 25)
  (lsp-ui-doc-header t)
  (lsp-ui-doc-include-signature t)
  (lsp-ui-doc-position 'top)
  (lsp-ui-doc-border (face-foreground 'default))
  (lsp-ui-sideline-enable t)
  (lsp-ui-sideline-ignore-duplicate t)
  (lsp-ui-sideline-show-code-actions t)
  :bind (:map lsp-ui-mode-map
         ([remap xref-find-definitions] . lsp-ui-peek-find-definitions)
         ([remap xref-find-references] . lsp-ui-peek-find-references)))
```

**OS**
macOS 11.0.1 using emacs 27.1 installed via `brew install --cask emacs`. 

Using `/sshx:fedora:/home/shanemcd/` into a Fedora 33 VM.

**Comment by hpdeifel** [comment]
I also have this on Fedora 33  when tramping into a debian docker container.

My current workaround is to set `*lsp-log-io*` to `t` while starting and set it back to `nil` after some time so the session stays responsive.

**Comment by ordicker** [comment]
Got the same thing with clang and linux [host-ubuntu remote-ubuntu(nvidia xavier)]

**Comment by zilti** [comment]
I have the same issue... `*lsp-log-io*` doesn't help. Where would that log the io to anyway? Maybe something in that log could help.

**Comment by ordicker** [comment]
That's my lsp-log
`
Command "clangd-10" is present on the path.
Command "clangd-10" is present on the path.
Found the following clients for /ssh:XXXXX@XXXXXXXX:/home/XXXX/workspace/pipeline/src/blur_exp.cpp: (server-id clangd-remote, priority 0)
The following clients were selected based on priority: (server-id clangd-remote, priority 0)
`
I don't find it extremely helpful :disappointed:  ,but again when you turn the login on the problem is solved.

**Comment by zilti** [comment]
Ah, it seems my mistake was to manually `(setq *lsp-log-io*)` in a scratch buffer. Adding it in my init file very high up helped. What an odd thing... Probably a timing issue, aka a Heisenbug?

**Comment by yyoncho** [comment]
I suspect the issue is that process-send-string is re-entrant and adding the logging slows down lsp-mode. Adding print statements in the beginning/ending of `lsp--send-no-wait` will help proving that.

**Comment by zilti** [comment]
Ah no, the `*lsp-log-io*` thing is not even good enough as a workaround. Even with it enabled, it manages to get past "starting..." once in about every five times, and usually after having to wait for a few minutes. And *if* it manages to get from starting to started, it locks up Emacs 75% of the time.

**Comment by md-arif-shaikh** [comment]
Experiencing the same issue where lsp-mode over tramp freezes emacs after few seconds.

**Comment by zilti** [comment]
It seems the freezing is "intentional", it freezes until it has analyzed the folder contents. For me it freezes Emacs for about one minute, before asking me if I want it to watch all files in my project.

**Comment by yyoncho** [comment]
@zilti most likely file watches over tramp are slow. I suggest turning them off by setting lsp-enable-file-watchers to nil

**Comment by zilti** [comment]
Yes, I assume that's the issue. I'm fine with it, there's worse things to happen than to have to wait a minute at the beginning of a work day before starting to code ^^ just wanted to explain the freezing. But turning it off is probably a good idea for most people.

**Comment by iostapyshyn** [comment]
I'm experiencing the same problem with clangd on emacs upstream. It does start occasionally and lsp-log-io doesn't  help

**Comment by antifuchs** [comment]
I've been trying to use a rust-analyzer LSP with `lsp-tramp-connection` for a few weeks now, and never went anywhere, so I decided to debug it. The problem seems to be that LSP tries to use stdio communication over a channel that is inherently crippled - ssh stdio with some semi-terminal emulation in the mix.

That starts with the stdout/stderr confusion (which lsp tries to work around by redirecting to a file), continues on to TTY problems and \n->\r translation (worked around with `stty raw`, which has its own stdout/stderr problems...) and then continues to weird buffering issues.

So, the solution for me ended up being to not use `lsp-tramp-connection` at all and use ssh's portforwarding instead.

Here's the connection definition that I'm using now for rust-analyzer, which works almost like a local connection (and is much faster than local, since the remote is a far more powerful machine than my laptop is):

``` elisp
(defun lsp-tramp-connection-over-ssh-port-forwarding (command)
  "Like lsp-tcp-connection, but uses SSH portforwarding."
  (list
   :connect (lambda (filter sentinel name environment-fn)
              (let* ((host "localhost")
                     (lsp-port (lsp--find-available-port host (cl-incf lsp--tcp-port)))
                     (command (with-parsed-tramp-file-name buffer-file-name nil
                                (message "[tcp/ssh hack] running LSP %s on %s / %s" command host localname)
                                (let* ((unix-socket (format "/tmp/lsp-ssh-portforward-%s.sock" lsp-port))
                                       (command (list
                                                 "ssh"
                                                 ;; "-vvv"
                                                 "-L" (format "%s:%s" lsp-port unix-socket)
                                                 host
                                                 "socat"
                                                 (format "unix-listen:%s" unix-socket)
                                                 (format "system:'\"cd %s && %s\"'" (file-name-directory localname) command)
                                                 )))
                                  (message "using local command %s" command)
                                  command)))
                     (final-command (if (consp command) command (list command)))
                     (_ (unless (executable-find (cl-first final-command))
                          (user-error (format "Couldn't find executable %s" (cl-first final-command)))))
                     (process-environment
                      (lsp--compute-process-environment environment-fn))
                     (proc (make-process :name name :connection-type 'pipe :coding 'no-conversion
                                         :command final-command :sentinel sentinel :stderr (format "*%s::stderr*" name) :noquery t))
                     (tcp-proc (progn
                                 (sleep-for 1) ; prevent a connection before SSH has run socat. Ugh.
                                 (lsp--open-network-stream host lsp-port (concat name "::tcp")))))

                ;; TODO: Same :noquery issue (see above)
                (set-process-query-on-exit-flag proc nil)
                (set-process-query-on-exit-flag tcp-proc nil)
                (set-process-filter tcp-proc filter)
                (cons tcp-proc proc)))
   :test? (lambda () t)))

(lsp-register-client
 (make-lsp-client :new-connection (lsp-tramp-connection-over-ssh-port-forwarding "rust-analyzer")
                  :major-modes '(rust-mode)
                  :initialization-options '((omitInitBuild . t)
                                            (cmdRun . t))
                  :notification-handlers (ht ("window/progress" 'lsp-clients--rust-window-progress))
                  :action-handlers (ht ("rls.run" 'lsp-rust--rls-run))
                  :remote? t
                  :server-id 'rust-analyzer-remote))
```

This code uses `socat` on the remote end to set up SSH forwarding such that the remote end is still an stdio-based language server (as that's the only way rust-analyzer runs at the moment), and connects through a unix domain socket (just so the remote connection doesn't occupy a potentially-taken local port). You could adjust this to listen on a TCP port, but this seemed safer to me.

Hope this is useful to somebody else. 

One other word of caution though, I have tried setting up the same connection function for the pyrite language server, but it stubbornly refuses to send responses back to emacs after opening a file, so something is wonky there still. Given that rust-analyzer works though, I'd say it is pyrite-language-server that is having a bad day, not the ssh forwarding setup.

**Comment by muirdm** [comment]
I also have startup issues over tramp that go away with a sleep. I'm going to assume it is the same issue (but note that hack language server crashes instead of hangs).

> I suspect the issue is that process-send-string is re-entrant and adding the logging slows down lsp-mode. Adding print statements in the beginning/ending of lsp--send-no-wait will help proving that.

I instrumented lsp--send-no-wait but it was not re-entrant. In fact, there was only a single call (the initialization message).

I changed my server command to start under strace so I could see exactly what the server was getting:

```
1926880 <... read resumed>"Content-Length: 2337\n\n\n\n{\"jsonrpc\":\"2.0\",\"method\":\"initialize\",\"params\":{\"processId\":null,\"rootPath\":\"XXX\",\"clientInfo\":{\"name\":\"emacs\",\"version\":\"GNUEmacs 28.0.50 (build 1, x86_64-apple-darwin20.5.0, NS appkit-2022.50 Version 11.4 (Build 20F71))\\n of 2021-06-15\"},\"rootUri\":\"file:///XXX\",\"capabilities\":{\"workspace\":{\"workspaceEdit\":{\"documentChanges\":true,\"resourceOperations\":[\"create\",\"rename\",\"delete\"]},\"applyEdit\":true,\"symbol\":{\"symbolKind\":{\"valueSet\":[1,2,3,4,5,6,7,8,9,10,11,12,13,14,15,16,17,18,19,20,21,22,23,24,25,26]}},\"executeCommand\":{\"dynamicRegistration\":false},\"workspaceFolders\":true,\"configuration\":true},\"textDocument\":{\"declaration\":{\"linkSupport\":true},\"definition\":{\"linkSupport\":true},\"implementation\":{\"linkSupport\":true},\"typeDefinition\":{\"linkSupport\":true},\"synchronization\":{\"willSave\":true,\"didSave\":true,\"willSaveWaitUntil\":true},\"documentSymbol\":{\"symbolKind\":{\"valueSet\":[1,2,3,4,5,6,7,8,9,10,11,12,13,14,15,16,17,18,19,20,21,22,23,24,25,26]},\"hierarchicalDocumentSymbolSupport\":true},\"formatting\":{\"dynamicRegistration\":true},\"rangeFormatting\":{\"dynamicRegistration\":true},\"rename\":{\"dynamicRegistration\":true,\"prepareSupport\":true},\"codeAction\":{\"dynamicRegistration\":true,\"isPreferredSupport\":true,\"codeActionLiteralSupport\":{\"codeActionKind\":{\"valueSet\":[\"\",\"quickfix\",\"refactor\",\"refactor.extract\",\"refactor.inline\",\"refactor.rewrite\",\"source\",\"source.organizeImports\"]}},\"resolveSupport\":{\"properties\":[\"edit\",\"command\"]},\"dataSupport\":true},\"completion\":{\"completionItem\":{\"snippetSupport\":true,\"documentationFormat\":[\"markdown\",\"plaintext\"],\"resolveAdditionalTextEditsSupport\":true,\"insertReplaceSupport\":true,\"resolveSupport\":{\"properties\":[\"documentation\",\"details\",\"additionalTextEdits\",\"command\"]},\"insertTextModeSupport\":{\"valueSet\":[1,2]}},\"contextSupport\":true},\"signatureHelp\":{\"signatureInformation\":{\"parameterInformation\":{\"labelOffsetSupport\":true}}},\"documentLink\":{\"dynamicRegistration\":true,\"tooltipSupport\":true},\"hover\":{\"contentFormat\":[\"markdown\",\"plaintext\"]},\"foldingRange\":{\"dynamicRegistration\":true},\"callHierarchy\":{\"dynamicRegistration\":false},\"publishDiagnostics\":{\"relatedInformation\":true,\"tagSupport\":{\"valueSet\":[1,2]},\"versionSupport\":true}},\"window\":{\"workDoneProgress\":true}},\"initializationOptions\":null,\"workDoneToken\":\"1\"},\"id\":1062}\n", 65536) = 2361
```

Notice the `\n\n\n\n` instead of `\r\n\r\n` after header. When I add a sleep before we send the first message the server gets `\r\n\r\n` properly.

I'm running Emacs 28.0.50 I built this week.

**Comment by muirdm** [comment]
I think we can fix it by echoing a string after the `stty` and waiting for it (so we can be sure the stty has finished before we start writing stuff to the process). Something like:

```diff
  diff --git a/lsp-mode.el b/lsp-mode.el
index e40775764..75e2ace99 100644
--- a/lsp-mode.el
+++ b/lsp-mode.el
@@ -6878,7 +6882,7 @@ returns the command to execute."
                           (process-name (generate-new-buffer-name name))
                           (wrapped-command (s-join
                                             " "
-                                            (append '("stty" "raw" ";")
+                                            (append '("stty" "raw" ";" "echo" "ready" ";")
                                                     final-command
                                                     (list
                                                      (concat "2>"
@@ -6891,6 +6895,14 @@ returns the command to execute."
                      (let ((proc (start-file-process-shell-command process-name
                                                                    (format "*%s*" process-name)
                                                                    wrapped-command)))
+
+                       (set-process-filter proc (lambda (proc input)
+                                                  (when (not (string= input "ready"))
+                                                    (lsp-warn "Unexpected startup message: %s" input))))
+
+                       (unless (accept-process-output proc 5)
+                         (lsp-warn "Didn't receive ready message on startup."))
+
                        (set-process-sentinel proc sentinel)
                        (set-process-filter proc filter)
                        (set-process-query-on-exit-flag proc nil)
```

**Comment by yyoncho** [comment]
@muirdm there is also this - https://github.com/emacs-lsp/lsp-mode/pull/2531 

Have you tested with latest version of tramp? Although this proposal might fix the issue I would like to avoid it because it seems more like a hack to avoid tramp bug.

**Comment by muirdm** [comment]
I'm using tramp 2.5.1-pre. #2531 looks like it should work instead of my patch - I'll try it later.

**Comment by yyoncho** [comment]
> But will people be able to use tramp 2.5 on stable Emacs?

I think that it is available via elpa.

**Comment by muirdm** [comment]
Apart from some minor bugs, #2531 seems to work great and gives a much better experience overall.

**Comment by zilti** [comment]
Any updates on this?

**Comment by sstepashka** [comment]
Any updates?

**Comment by manuel-arguelles** [comment]
In my case it starts clangd-12 in the container and starts indexing just file, but emacs is locked... cpu usage is not at 100% but it's totally unresponsive. (emacs 27.1) tramp 2.5

**Comment by Blade6570** [comment]
Any updates on it? One year has passed and not able to find any other discussion on it.

**Comment by KJordanBerg** [comment]
I'm also interested to know if there are any updates. I'd really like to use LSP-mode over Tramp.

**Comment by TMueller83** [comment]
I'm affected by the same problem too. Would be happy for a fix.

**Comment by b-spencer** [comment]
FWIW, I managed to catch this in a locally rebuilt version of the stock Debian 10 backport of `emacs-lucid` 27.1 with the debugger.  I am not familiar with emacs internals, but I was able to break execution at many places and it looked like it was evaluating Lisp over and over again forever.  In other words, if the C core of Emacs was in an infinite loop, it sure wasn't a tight one.  It felt like it could have been an infinite loop in Lisp, or perhaps an unending series of Lisp events to process.  At one point I caught it deep in the garbage collector, but the rest of the time it was churning through byte code at many layers of C.

**Comment by albusshin** [comment]
I've been using @antifuchs 's code above until recently I upgraded all the packages in my doom emacs, and it stopped working with the error of `wrong-number-of-arguments`.

This was due to the number of argument changed in commit `59eeba73a6e1064d10efefa9a8ba0de20c938591`. 
The following modified version should work:

```
(defun lsp-tramp-connection-over-ssh-port-forwarding (command)
  "Like lsp-tcp-connection, but uses SSH portforwarding."
  (list
   :connect (lambda (filter sentinel name environment-fn _workspace)
              (let* ((host "localhost")
                     (lsp-port (lsp--find-available-port host (cl-incf lsp--tcp-port)))
                     (command (with-parsed-tramp-file-name buffer-file-name nil
                                (message "[tcp/ssh hack] running LSP %s on %s / %s" command host localname)
                                (let* ((unix-socket (format "/tmp/lsp-ssh-portforward-%s.sock" lsp-port))
                                       (command (list
                                                 "ssh"
                                                 ;; "-vvv"
                                                 "-L" (format "%s:%s" lsp-port unix-socket)
                                                 host
                                                 "socat"
                                                 (format "unix-listen:%s" unix-socket)
                                                 (format "system:'\"cd %s && %s\"'" (file-name-directory localname) command)
                                                 )))
                                  (message "using local command %s" command)
                                  command)))
                     (final-command (if (consp command) command (list command)))
                     (_ (unless (executable-find (cl-first final-command))
                          (user-error (format "Couldn't find executable %s" (cl-first final-command)))))
                     (process-environment
                      (lsp--compute-process-environment environment-fn))
                     (proc (make-process :name name :connection-type 'pipe :coding 'no-conversion
                                         :command final-command :sentinel sentinel :stderr (format "*%s::stderr*" name) :noquery t))
                     (tcp-proc (progn
                                 (sleep-for 1) ; prevent a connection before SSH has run socat. Ugh.
                                 (lsp--open-network-stream host lsp-port (concat name "::tcp")))))

                ;; TODO: Same :noquery issue (see above)
                (set-process-query-on-exit-flag proc nil)
                (set-process-query-on-exit-flag tcp-proc nil)
                (set-process-filter tcp-proc filter)
                (cons tcp-proc proc)))
   :test? (lambda () t)))

(lsp-register-client
 (make-lsp-client :new-connection (lsp-tramp-connection-over-ssh-port-forwarding "rust-analyzer")
                  :major-modes '(rust-mode)
                  :initialization-options '((omitInitBuild . t)
                                            (cmdRun . t))
                  :notification-handlers (ht ("window/progress" 'lsp-clients--rust-window-progress))
                  :action-handlers (ht ("rls.run" 'lsp-rust--rls-run))
                  :remote? t
                  :server-id 'rust-analyzer-remote))
```

**Comment by zilti** [comment]
Because the workaround is not actually working. And at least for me it also happened in Spacemacs.

**Comment by seokbeomKim** [comment]
@zilti I see ....  then I need to find an alternative for working on remote sessions. Thank you anyway.

**Comment by rudihorn** [comment]
I have been facing the same issue, and the ssh workaround has not been ideal, and so I have been trying to diagnose this issue a little.

Essentialy the symptoms I see are that some bytes get lost. When replacing my lsp command with something that essentially does `tee <logfile> | lsp`, the logfile starts with:

```
nitialize","params":{...
```

This seems to indicate that it isn't a whole packet getting lost, just a few bytes. @yyoncho you mentioned it may be because the call is re-entrant, but to me it feels like this is a symptom of something else.

My understanding of what is going on is something like this:

1. Tramp invokes ssh something like `ssh -q ... -o RemoteCommand=/bin/sh  -i <host>`
2. Tramp sets up the environment by invoking `"cd <dir> && %s exec %s %s env %s <command>"` this includes the `INSIDE_EMACS` what shows up in the `.bash_history`. This is invoked using `(tramp-send-command v command nil t)`. Importantly `NOOUTPUT=t` here, so it does not wait for completion.
3. The command invoked by tramp in 2.) is something like `/bin/sh -c stty raw > /dev/null; <lsp>`
4. `lsp-mode` more or less immediately invokes `lsp-process-send` with the initialize message, which in turn uses `(process-send-string <process> <initialize>)`

Somewhere along this path, a few bytes from the initialize message are taken before they reach the `tee` command, and it is a bit difficult to diagnose where. I'm wondering if incorrect configuration of the initial shell could mean that the shell is reading bytes beyond the initial command, and maybe this could be an issue with TRAMP in general.

A workaround for all of this could potentially be for `lsp-mode` to rather construct a command such as:

```
/bin/sh -c stty raw > /dev/null; echo READY; <lsp>`
```

And then only send the initial command after receiving the `READY`. I think at that point it should be safe to communicate with the stdin.

An even quicker workaround is to patch up `lsp-stdio-connection` by adding a `sleep-for 0.5` or something:

```
                                     :stderr stderr-buf
                                     :noquery t
                                     :file-handler t))
+                              (wait (sleep-for 0.5)))
                         (set-process-query-on-exit-flag proc nil)
                         (set-process-query-on-exit-flag (get-buffer-process stderr-buf) nil)
                         (with-current-buffer (get-buffer stderr-buf)
                           ;; Make the *NAME::stderr* buffer buffer-read-only, q to bury, etc.
```

Any thoughts on this? It would be really nice if this could be patched properly.

EDIT: some of this is quite likely also related to my usage of `sshx:` and its usage of `-t` instead of `-T`.

**Comment by rudihorn** [comment]
A bit more poking around and it seems like part of what triggers this issue of stdin contents getting lost is the use of `TERM=dumb INSIDE_EMACS=foo`. On the system I am testing with `/bin/sh` is provided by bash, which is slightly different to e.g. Ubuntu where dash  is used instead. This ends up triggering bash to use `no_read_line` (https://github.com/bminor/bash/blob/bash-4.4/shell.c#L610) which is roughly equivalent to using bash with the `--noediting` mode. The simplest reproducer I could find was to create a text file with the contents:

```
exec env TERM=dumb INSIDE_EMACS=ABC /bin/bash -i
exec env /bin/bash -i -c cat\ -
Lorem ipsum dolor sit amet, consetetur sadipscing elitr, sed diam nonumy eirmod tempor invidunt ut labore et dolore magna aliquyam erat, sed diam voluptua. At vero eos et accusam et justo duo dolores et ea rebum. Stet clita kasd gubergren, no sea takimata sanctus est Lorem ipsum dolor sit amet. Lorem ipsum dolor sit amet, consetetur sadipscing elitr, sed diam nonumy eirmod tempor invidunt ut labore et dolore magna aliquyam erat, sed diam voluptua. At vero eos et accusam et justo duo dolores et ea rebum. Stet clita kasd gubergren, no sea takimata sanctus est Lorem ipsum dolor sit amet.
Lorem ipsum dolor sit amet, consetetur sadipscing elitr, sed diam nonumy eirmod tempor invidunt ut labore et dolore magna aliquyam erat, sed diam voluptua. At vero eos et accusam et justo duo dolores et ea rebum. Stet clita kasd gubergren, no sea takimata sanctus est Lorem ipsum dolor sit amet. Lorem ipsum dolor sit amet, consetetur sadipscing elitr, sed diam nonumy eirmod tempor invidunt ut labore et dolore magna aliquyam erat, sed diam voluptua. At vero eos et accusam et justo duo dolores et ea rebum. Stet clita kasd gubergren, no sea takimata sanctus est Lorem ipsum dolor sit amet.
Lorem ipsum dolor sit amet, consetetur sadipscing elitr, sed diam nonumy eirmod tempor invidunt ut labore et dolore magna aliquyam erat, sed diam voluptua. At vero eos et accusam et justo duo dolores et ea rebum. Stet clita kasd gubergren, no sea takimata sanctus est Lorem ipsum dolor sit amet. Lorem ipsum dolor sit amet, consetetur sadipscing elitr, sed diam nonumy eirmod tempor invidunt ut labore et dolore magna aliquyam erat, sed diam voluptua. At vero eos et accusam et justo duo dolores et ea rebum. Stet clita kasd gubergren, no sea takimata sanctus est Lorem ipsum dolor sit amet.
```

Then invoke `cat cmds | /bin/bash -i`, and some of the input gets lost. Removing `TERM` or `INSIDE_EMACS` fixes it here.

From within tramps it was fairly easy to replicate the same issue. However there, just the presence of `INSIDE_EMACS` was sufficient to break it. It was possible to fix the issue by invoking the initial ssh /bin/sh command with `--noediting`, but this seems to make things worse when using the simple "pipe to bash" example (though I'm using a different version here). According to the bash source code, all of the shells should be interactive, so I don't think bash should be entering the buffered mode. I'm not yet sure why input gets lost when readline is disabled.

Edit: It seems bash without readline just [buffers](https://github.com/bminor/bash/blob/bash-4.4/input.c#L75C1-L75C19) the input stream when not using readline, with no method of ensuring the contents of `localbuf` are put back into stdin. I'm not sure I see much of an alternative to just flushing out this old data by sending a chunk of 128 bytes unless bash is patched.
Edit2: It seems `clangd` doesn't mind an additional `(process-send-string proc (make-string 128 ?\n))`, but e.g. `pylsp` does.
Edit3: Setting the remote shell to `/usr/bin/dash` by invoking `(add-to-list 'tramp-connection-properties (list (regexp-quote "sshx:host:") "remote-shell" "/usr/bin/dash"))` and then replacing `shell-file-name` with `/usr/bin/bash` in `lsp-mode.el` seems to fix it. I'm not sure if there is a better way to replace `shell-file-name`.

**Comment by kurahaupo** [comment]
@rudihorn 

(I'm here after reading a bug report on `bug-bash`; I don't know anything about what `lsp` does or how it works.)

The invocation `ssh -q … -o RemoteCommand=/bin/sh  -i $host` is unusual to say the least. I'm going to assume that `-i` is part of the remote command, since `ssh -i` would consume the following argument as the name of an identity file, and ssh would then complain that you hadn't provided a _host_ argument.

Forcing the shell into interactive mode using `-i` is only best-effort if stdin is not a tty.

Without readline, the shell relies on the kernel's tty driver to split up the input into lines. However if the stdin of `ssh` isn't a tty then it won't ask `sshd` to provide a tty on the remote end unless you use `ssh -tt`. (And the `stty` command is useless, for the same reason.)

Without a tty, the kernel simply hands over the maximum number of bytes without regard for line boundaries, which consumes some of the input.

Question: does `lsp` need its stdin or stdout to be a tty?
And why is the shell in interactive mode?
(Are you trying to mimic a human?)

Consider either:
1. Sending the command to run as part of the `ssh` invocation, with the input on stdin; run:
   ```
   ssh "$host" 'stty raw > /dev/null; echo READY; lsp'
   ```
   with whatever input you normally would use
2. Sending the script with its input as a heredoc:
   run
   ```
   ssh "$host" sh
   ```
   with input 
   ```
   echo READY
   lsp <<\EndOfInput
   …input here…
   EndOfInput
   ```
3. Install a short script on the target host
   ```
   #!/bin/sh
   echo READY
   lsp
   ```
   and simply invoke that:
   ```
   ssh "$host" /path/to/script
   ```

-Martin

PS: add `-tt` and then `stty raw` only if necessary.

**Comment by rudihorn** [comment]
@kurahaupo 

thank you very much for your response, it is definitely helps clarify a lot.

For background: LSP relies on tramp to establish the connection. Tramp is the general purpose tool used to access and edit remote files, as well as invoke remote programs. It is unfortunately used for almost anything, and is the reason why I'm not sure how much can be changed. However that does open up the question if LSP should avoid using tramp, though it would be nice if there was a general purpose way to use tramp.

The "good" news is that ssh does indeed use `-tt`, and I wasn't aware that the kernel tty driver splits up lines. However this in theory should fix the problem we have: bash should only read until the newline, and leave the rest for the program it invokes.

It turns out that this is actually causing the exact opposite problem that I had expected. Consider the following two invocations:

```
$ ssh -q -o ControlMaster=no -o ControlPath=none -e none -tt -o "RemoteCommand=/bin/sh --noediting -i" <host> <<EOF
strace -e read cat -
hello
world
EOF
...
read(0, "hello\n", ...)
hello
read(0, "world\n", ...)
world
```

```
$ ssh -q -o ControlMaster=no -o ControlPath=none -e none -tt -o "RemoteCommand=/bin/sh -i" <host> <<EOF
strace -e read cat -
hello
world
EOF
...
read(0, "hello\nword\n", ...)
hello
world
```

Here invoking the application with readline actually circumvents the kernel's tty driver behaviour for the called application, and lets the `cat -` application read two lines of text at once. Without readline (i.e. adding the `--noediting` flag or adding `env INSIDE_EMACS=ABC TERM=dumb`, the application works as expected.

Back to the lsp problem: The lsp is invoked in a shell in a shell over ssh. The first shell works fine, but is invoked with readline behaviour. The second shell is invoked without readline behaviour, and the first shell circumvents the second shells line feed behaviour. One thing I had noticed, is that adding `--noediting` to the first shell fixes the problem in the second shell.

I will need to dig into the bash sources to investigate why/how readline alters the input fd of the child process. I suspect it might create a separate pipe for the subprocess or something, but may also just change some configuration.

Regarding your proposed solutions:

1. Directly invoking the correct command remotely is an option, but it only fixes LSP and may be tricky to get right in a general way. It also prevents lsp-mode from reusing much of the tramp logic and configuration settings for connection setup. Tramp also tries to offer the opportunity to transparently spawn a process on the remote host by just invoking `(make-process ...)` with some magic that makes the process run on the remote host, only depending on what the current buffer is connected to.
2. A heredoc doesn't work because the LSP process is an interactive daemon process, where commands are sent back and forth. So the message sent being lost is only the first of many.
3. Suffers from the same issue as 1.)

Another potential workaround for the moment could be to permit the remote host to set `AcceptEnv INSIDE_EMACS TERM` on the remote host sshd, then add `-o "SetEnv=INSIDE_EMACS=ABC TERM=dumb"` to the tramp ssh control master options (or to the remote host config). That would force the first shell to not be readline, which seems to solve the whole issue. I will try experimenting with this in addition to looking why readline changes bash behaviour.

Edit: Given how old this issue is, are any of the others who were active still interested in a fix here? If so could you confirm if you have been facing the exact same issue caused by a similar setup (i.e. remote bash instead of other shell).
Edit2: It seems that in general bash just disables `ICANON` in `prepare_terminal_settings` (see man page `termios(3)`), and so the tty line-by-line behaviour is disabled during the execution of all commands, and restored on the way back.

**Comment by OrbitalHustler** [comment]
I'm having the same issue. Looking at the CPU call trace I noticed this which makes me thinks this is a re-entrant problem:

```
       22873  85% - timer-event-handler
       22873  85%  - apply
       22858  84%   - lsp--on-idle
       22858  84%    - lsp-diagnostics--flycheck-buffer
       22858  84%     - flycheck-buffer
       22854  84%      - flycheck-get-checker-for-buffer
       22854  84%       - flycheck-may-use-checker
       22854  84%        - #<byte-code-function 553>
       22854  84%         - flycheck-compute-working-directory
       22854  84%          - tramp-file-name-handler
       22854  84%           - apply
       22854  84%            - tramp-sh-file-name-handler
       22794  84%             - tramp-sh-handle-set-visited-file-modtime
       22794  84%              - tramp-file-name-handler
       22794  84%               - apply
       22794  84%                - tramp-sh-file-name-handler
       22794  84%                 - tramp-sh-handle-file-attributes
       22794  84%                  - tramp-do-file-attributes-with-stat
       22794  84%                   - tramp-send-command-and-read
       22794  84%                    - tramp-barf-unless-okay
       22794  84%                     - tramp-send-command-and-check
=======================
       22794  84%                      - tramp-send-command <<<<<<<<<<<<<<<<<<<<<<<<<<<<<<
=======================
       22794  84%                       - tramp-wait-for-output
=======================
       22794  84%                        - tramp-wait-for-regexp <<<<<<<<<<<<<<<<<<<<<<<<<<<<<<
=======================
       22766  84%                         - sit-for
       22766  84%                          - timer-event-handler
       22766  84%                           - apply
       22611  84%                            - #<byte-code-function 9D9>
       22611  84%                             - lsp--init-if-visible
       22611  84%                              - lsp
       22611  84%                               - apply
       22611  84%                                - lsp@magit-find-file
       22611  84%                                 - #<native-comp-function lsp>
       22479  83%                                  - lsp-find-session-folder
       22367  83%                                   - lsp-f-same?
       22367  83%                                    - f-exists?
       22367  83%                                     - tramp-file-name-handler
       22362  83%                                      - apply
       22362  83%                                       - tramp-sh-file-name-handler
       22362  83%                                        - tramp-sh-handle-file-exists-p
       22212  82%                                         - file-symlink-p
       22212  82%                                          - tramp-file-name-handler
       22212  82%                                           - apply
       22212  82%                                            - tramp-sh-file-name-handler
       22212  82%                                             - tramp-handle-file-symlink-p
       22204  82%                                              - file-attributes
       22204  82%                                               - tramp-file-name-handler
       22204  82%                                                - apply
       22204  82%                                                 - tramp-sh-file-name-handler
       22204  82%                                                  - tramp-sh-handle-file-attributes
       22204  82%                                                   - tramp-do-file-attributes-with-stat
       22204  82%                                                    - tramp-send-command-and-read
       22204  82%                                                     - tramp-barf-unless-okay
       22204  82%                                                      - tramp-send-command-and-check
=======================
       22204  82%                                                       - tramp-send-command <<<<<<<<<<<<<<<<<<<<<<<<<<<<<<
=======================
       22204  82%                                                        - tramp-wait-for-output
=======================
       22204  82%                                                         - tramp-wait-for-regexp <<<<<<<<<<<<<<<<<<<<<<<<<<<<<<
=======================
       12364  45%                                                          - tramp-accept-process-output
        8569  31%                                                             accept-process-output
```

We can see that `lsp--on-idle` triggers a flycheck call, that ends up triggering a `tramp-send-command` call, which calls `tramp-wait-for-regexp` which will wait with `sit-for`. 

Then, while it waits for the output of the command it just sent by ssh, a timer kicks in, that starts `lsp--init-if-visible` which also ends up calling `tramp-send-command`. 

So another command is sent over ssh before the first `tramp-wait-for-regexp` finished. 

When I look at the tramp log when this happens, I can see the issue:

```
16:28:27.843608 tramp-send-command (6) # tramp_stat_file_attributes /path/utils 2>/dev/null; echo tramp_exit_status $?
16:28:27.883875 tramp-do-file-attributes-with-stat (5) # file attributes with stat: /path/utils
16:28:27.884512 tramp-send-command (6) # tramp_stat_file_attributes /path/utils 2>/dev/null; echo tramp_exit_status $?
16:28:27.884825 tramp-wait-for-regexp (6) #
(("‘/path/utils’") 4 ("root" . 0) ("root" . 0) 1736807726 1759106458 1759106458 4096 "drwxr-xr-x" t 794856 -1)
tramp_exit_status 0
///4406293e4a51c89b848e26ac4acb177e#$
16:28:27.885191 tramp-send-command (6) # test -e /path/utils 2>/dev/null; echo tramp_exit_status $?
16:28:27.905910 tramp-wait-for-regexp (6) #
(("‘/path/utils’") 4 ("root" . 0) ("root" . 0) 1736807726 1759106458 1759106458 4096 "drwxr-xr-x" t 794856 -1)
tramp_exit_status 0
///4406293e4a51c89b848e26ac4acb177e#$tramp_exit_status 0
///4406293e4a51c89b848e26ac4acb177e#$
```

We see that two `tramp_stat_file_attributes` are sent in a row. Then we receive the output of the first, then we send a `test -e /path/utils` command, and the output of that command gets merged with the output of the second `tramp_stat_file_attributes`:

```
16:28:27.905910 tramp-wait-for-regexp (6) #
(("‘/path/utils’") 4 ("root" . 0) ("root" . 0) 1736807726 1759106458 1759106458 4096 "drwxr-xr-x" t 794856 -1)
tramp_exit_status 0
///4406293e4a51c89b848e26ac4acb177e#$tramp_exit_status 0
///4406293e4a51c89b848e26ac4acb177e#$
```

I believe this explain why, once all the calls have been processed, tramp hangs: it thinks one of the command hasn't received its output yet

**Comment by OrbitalHustler** [comment]
I believe the issue is caused by `lsp--init-if-visible` being called in a timer, which ends up calling tramp-send-command

      It is in general not a good idea to invoke synchronous Tramp operations
      in asynchronously running functions like timers, process filters and
      sentinels, event handlers. The Tramp manual mentions this in the FAQ
      section.

https://lists.gnu.org/archive/html/tramp-devel/2021-04/msg00036.html


The following elisp seems to fix the problem for me:

```elisp
;;; Track currently running tramp-send-command
(defvar lsp--tramp-connection-depths (make-hash-table :test 'equal)
  "Hash table tracking depth of tramp operations per connection.
    Keys are connection identifiers, values are operation depth counts.")

(defun lsp--tramp-connection-key (vec)
  "Get unique connection identifier from tramp VEC."
  (when vec
    (format "%s|%s|%s|%s"
            (or (tramp-file-name-method vec) "")
            (or (tramp-file-name-user vec) "")
            (or (tramp-file-name-host vec) "")
            (or (tramp-file-name-port vec) ""))))

(defun lsp--tramp-connection-key-from-file (file-name)
  "Get connection identifier from FILE-NAME."
  (when (and file-name (file-remote-p file-name))
    (let ((vec (tramp-dissect-file-name file-name)))
      (lsp--tramp-connection-key vec))))

(defun lsp--tramp-track-operations-advice (orig-fun vec command &optional neveropen nooutput)
  "Track depth of tramp operations per connection."
  (let* ((conn (lsp--tramp-connection-key vec))
         (current-depth (if conn (gethash conn lsp--tramp-connection-depths 0) 0)))

    (when conn
      ;; Increment depth before operation
      (puthash conn (1+ current-depth) lsp--tramp-connection-depths))

    (unwind-protect
        (funcall orig-fun vec command neveropen nooutput)

      ;; Decrement depth after operation (even on error)
      (when conn
        (let ((new-depth (1- (gethash conn lsp--tramp-connection-depths 1))))
          (if (<= new-depth 0)
              (remhash conn lsp--tramp-connection-depths)
            (puthash conn new-depth lsp--tramp-connection-depths)))))))

(advice-add 'tramp-send-command :around #'lsp--tramp-track-operations-advice)

(defun lsp--tramp-connection-busy-p (file-name)
  "Check if the tramp connection for FILE-NAME is currently busy.
    Returns non-nil if there's an active tramp operation on this connection."
  (let ((conn (lsp--tramp-connection-key-from-file file-name)))
    (and conn (> (gethash conn lsp--tramp-connection-depths 0) 0))))

(defvar-local lsp--init-if-visible-in-progress nil
  "Non-nil when lsp--init-if-visible is currently executing.")

;;; Reschedule lsp--init-if-visible if the tramp connection is busy running a command
(with-eval-after-load 'lsp-mode

  (defun lsp--init-if-visible-with-tramp-guard (orig-fun)
    "Advice for lsp--init-if-visible to handle busy tramp connections."
    (let ((file-name (buffer-file-name))
          (buffer (current-buffer)))
      (cond
       ;; Already running in this buffer - just abort, don't reschedule
       (lsp--init-if-visible-in-progress
        nil)

       ;; Tramp connection busy - reschedule and abort
       ((and file-name
             (file-remote-p file-name)
             (lsp--tramp-connection-busy-p file-name))
        (run-with-idle-timer 0.1 nil
                             (lambda ()
                               (when (buffer-live-p buffer)
                                 (with-current-buffer buffer
                                   (lsp--init-if-visible)))))
        nil)

       ;; All clear - proceed with initialization
       (t
        (setq lsp--init-if-visible-in-progress t)
        (unwind-protect
            (funcall orig-fun)
          (setq lsp--init-if-visible-in-progress nil))))))

  (advice-add 'lsp--init-if-visible :around #'lsp--init-if-visible-with-tramp-guard))

```

The fix is:
- Add an advice around all `tramp-send-command` that will help keep track of which tramp connection is currently busy running a command and waiting for the output of that command
- Add an advice around `lsp--init-if-visible`:
  - It will prevent re-entrant calls of `lsp--init-if-visible` when it's already running, as I saw it happen
  - If the file is remote, and the tramp connection of that file is currently running a command (we can know that thanks to the first advice), reschedule `lsp--init-if-visible` 0.1sec later

Could someone test this and give feedback?

**Comment by rudihorn** [comment]
@OrbitalHustler sorry for the slight delay in response.

I have been trying your workaround for a bit. Interestingly it does seem to also improve on the startup of my LSP, even though I'm pretty sure the bash stdin issue continues to exist.

I think it does improve the reentrant tramp calls, but I'm not sure they entirely go away. Or at least I got the impression that the `lsp-tramp-connection-over-ssh-port-forwarding` workaround was less problematic in this regard. I still need to look over your code in a bit more detail to understand what is going on. I'm also still a bit confused, because it seems like this is breaking the tramp connections used for other things, and there should be (in my understanding) two tramp connections: One for the LSP (which does not use the control master), and one for everything else such as fetching the file (with control master). If these are all sharing a tramp connection, is there a chance your workaround only prevents the LSP from tripping up over itself, but it does not prevent other modules from sending requests? Though these should be different sessions because the LSP one does not permit shell usage.

**Comment by OrbitalHustler** [comment]
@rudihorn You're correct, my workaround targets lsp specifically:
- it prevents `lsp--init-if-visible` from running again while it's already running in the buffer (this happens when the first `lsp--init-if-visible` is waiting for a tramp request to finish, and a timer starts it a second time). It just aborts the second call
- it stops `lsp--init-if-visible` if it's about to run on a buffer for which tramp is already sending a command & waiting for the response. In that case it gets rescheduled to run again in 0.1 sec

But other packages can still use tramp in problematic ways. I see sometimes message about forbidden tramp reentrant calls, for example. But at least with this workaround, I can open a remote python buffer and have lsp automatically start in it without freezing.

> I'm also still a bit confused, because it seems like this is breaking the tramp connections used for other things, and there should be (in my understanding) two tramp connections: One for the LSP (which does not use the control master), and one for everything else such as fetching the file (with control master). If these are all sharing a tramp connection, is there a chance your workaround only prevents the LSP from tripping up over itself, but it does not prevent other modules from sending requests? Though these should be different sessions because the LSP one does not permit shell usage.

About the two connections: my understanding is that `lsp--init-if-visible` itself doesn't communicate with the LSP server; it does file operations (checking directories, finding project roots, etc.) using the regular file operations tramp connection. That's the same connection flycheck uses, which is why they collide.

Anything that is not `lsp--init-if-visible` should be completely unaffected by my workaround. All tramp operations will cause the `lsp--tramp-connection-depths` to be modified, but it's just bookkeeping, it doesn't block or impact anything else. The advice `lsp--init-if-visible-with-tramp-guard` is the thing that is altering lsp & tramp interaction
